Source/WebCore/platform/network/curl/CurlMultipartHandle.h

```diff
@@ -41,37 +41,42 @@ class CurlMultipartHandle {
 public:
     static std::unique_ptr<CurlMultipartHandle> createIfNeeded(CurlMultipartHandleClient&, const CurlResponse&);
 
-    CurlMultipartHandle(CurlMultipartHandleClient&, const String&);
+    CurlMultipartHandle(CurlMultipartHandleClient&, String&&);
     ~CurlMultipartHandle() { }
 
-    void didReceiveData(const SharedBuffer&);
-    void didComplete();
+    void didReceiveMessage(const SharedBuffer&);
+    void didCompleteMessage();
+
+    void completeHeaderProcessing();
+
+    bool completed() { return m_didCompleteMessage; }
+    bool hasError() const { return m_hasError; }
 
 private:
     enum class State {
-        CheckBoundary,
-        InBoundary,
+        FindBoundaryStart,
         InHeader,
-        InContent,
-        EndBoundary,
+        WaitingForHeaderProcessing,
+        InBody,
+        WaitingForTerminate,
+        Terminating,
         End
     };
 
-    static std::optional<String> extractBoundary(const CurlResponse&);
-    static std::optional<String> extractBoundaryFromContentType(const String&);
-
     bool processContent();
-    bool checkForBoundary(size_t& boundaryStartPosition, size_t& lastPartialMatchPosition);
-    size_t matchedLength(const uint8_t* data);
+    std::optional<bool> findBoundary(size_t& matchPos, size_t& endPos, bool hasCloseDelimiter);
     bool parseHeadersIfPossible();
+    bool shouldTerminateIfNeeded();
 
     CurlMultipartHandleClient& m_client;
 
     String m_boundary;
     Vector<uint8_t> m_buffer;
```

Comment:
Using Vector is inefficient here. SharedBuffer is better. But, it might make the code too complicated. Just an idea.

Reply:
I agree. However, using SharedBuffer makes the processing very complicated, so I would like to enable supporting multipart in the current processing. And if the performance is really bad, we should consider moving to SharedBuffer.

Reply:
It's reasonable.